### PR TITLE
Ban split of list-partitioned sub-partitions

### PR DESCRIPTION
**Pull request by melanieplageman**
Currently, GPDB does not support splitting a multi-level partition when
the leaf partition is partitioned by list. However, in many cases, we
did not correctly error out and instead crashed. This commit detects a
query attempting to split a list partitioned sub-partition and correctly
errors out.

Signed-off-by: Sam Dash <sdash@pivotal.io>

### CODE REVIEW

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -2981,6 +2982,54 @@ ATController(Relation rel, List *cmds, bool recurse)
 
 }
 
+static void
+prepSplitCmd(Relation rel, PgPartRule *prule, bool is_at)
```
Comment (hsyuan) [inline src/backend/commands/tablecmds.c]: Can you add comment for this function?

File: src/backend/commands/tablecmds.c [hunk 2]
```diff
@@ -2981,6 +2982,54 @@ ATController(Relation rel, List *cmds, bool recurse)
...
+		prule->pNode->part->parkind == 'r' &&
+		is_at)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("AT clause cannot be used when splitting "
+						"a default RANGE partition")));
+	}
+	else if (prule->pNode->part->parkind == 'l' && pNode->default_part)
+		if (pNode->default_part->children)
```
Comment (hsyuan) [inline src/backend/commands/tablecmds.c]: add a pair of `{}` outside of this block

Comment (hsyuan) [reply src/backend/commands/tablecmds.c]: why not add this condition to the line above it?

File: src/backend/commands/tablecmds.c [hunk 3]
```diff
@@ -12752,6 +12768,7 @@ ATPExecPartAlter(List **wqueue, AlteredTableInfo *tab, Relation rel,
 	AlterPartitionCmd 	*pc2		   = NULL;
 	bool				 bPartitionCmd = true;	/* true if a "partition" cmd */
 	Relation			 rel2		   = rel;
+	bool				prepCmd		= false;
```
Comment (hsyuan) [inline src/backend/commands/tablecmds.c]: comment for this var, please

File: src/backend/commands/tablecmds.c [hunk 4]
```diff
@@ -12837,6 +12857,18 @@ ATPExecPartAlter(List **wqueue, AlteredTableInfo *tab, Relation rel,
 			pid2->location  = -1;
 
 			pc2->partid = (Node *)pid2;
+
+			if (prepCmd)
+			{
+				PgPartRule			*prule1	= NULL;
+				bool is_at = true;
```
Comment (hsyuan) [inline src/backend/commands/tablecmds.c]: comment for vars

File: src/backend/commands/tablecmds.c [hunk 5]
```diff
@@ -12837,6 +12857,18 @@ ATPExecPartAlter(List **wqueue, AlteredTableInfo *tab, Relation rel,
...
+				prule1 = get_part_rule(rel, pid2, true, true, NULL, false);
+
+				if (linitial((List *)pc2->arg1))
```
Comment (hsyuan) [inline src/backend/commands/tablecmds.c]: why?

File: src/backend/commands/tablecmds.c [hunk 6]
```diff
@@ -2981,6 +2982,54 @@ ATController(Relation rel, List *cmds, bool recurse)
...
+						"a default RANGE partition")));
+	}
+	else if (prule->pNode->part->parkind == 'l' && pNode->default_part)
+		if (pNode->default_part->children)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_GP_FEATURE_NOT_SUPPORTED),
+						errmsg("split partition is not "
+								"currently supported when the "
+								"lowest level is list partitioned")));
```
Comment (kavinderd) [inline src/backend/commands/tablecmds.c]: This message can be a little more concise: "SPLIT PARITITON is not supported for leaf list partitions", or something like that

File: src/backend/commands/tablecmds.c [hunk 7]
```diff
@@ -2981,6 +2982,54 @@ ATController(Relation rel, List *cmds, bool recurse)
 
 }
 
+static void
+prepSplitCmd(Relation rel, PgPartRule *prule, bool is_at)
+{
+	PartitionNode		*pNode = NULL;
+	pNode = RelationBuildPartitionDesc(rel, false);
```
Comment (dhanashreek89) [inline src/backend/commands/tablecmds.c]: These two lines can be combined.

File: src/backend/commands/tablecmds.c (hunk 2 above)
Comment (dhanashreek89) [inline src/backend/commands/tablecmds.c]: Please check if pNode is not null before accessing:
`if (pNode && pNode->default_part->children)`

File: src/backend/commands/tablecmds.c [hunk 8]
```diff
@@ -12773,11 +12790,14 @@ ATPExecPartAlter(List **wqueue, AlteredTableInfo *tab, Relation rel,
 
 	switch (atc->subtype)
 	{
+		case AT_PartSplit:				/* Split */
+		{
+			prepCmd = true;
+		}
```
Comment (dhanashreek89) [inline src/backend/commands/tablecmds.c]: Do we need a break here ?

File: src/backend/commands/tablecmds.c (hunk 5 above)
Comment (dhanashreek89) [reply src/backend/commands/tablecmds.c]: a comment would be good here

**Comment by ashwinstar** [comment]
Does this fix issue https://github.com/greenplum-db/gpdb/issues/2032 as well then ?

**Comment by sambitesh** [comment]
@ashwinstar Yes it does fix the issue.

File: src/backend/commands/tablecmds.c [hunk 9]
```diff
@@ -2981,6 +2982,66 @@ ATController(Relation rel, List *cmds, bool recurse)
...
+	}
+
+	if (prule->topRule->parisdefault &&
+		prule->pNode->part->parkind == 'r' &&
+		is_at)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("AT clause cannot be used when splitting "
+						"a default RANGE partition")));
```
Comment (danielgustafsson) [inline src/backend/commands/tablecmds.c]: It's common practice in upstream, and I believe it should be for GPDB as well, to not linebreak error messages but instead let them exceed the recommended line length. It makes it easier to grep the code for a specific error message. This has not been done historically for GPDB, but I think going forward we should keep them un-broken.

File: src/backend/commands/tablecmds.c [hunk 10]
```diff
@@ -2981,6 +2982,66 @@ ATController(Relation rel, List *cmds, bool recurse)
...
+	}
+	else if ((prule->pNode->part->parkind == 'l') && (pNode !=NULL && pNode->default_part)
+			 && (pNode->default_part->children))
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_GP_FEATURE_NOT_SUPPORTED),
+						errmsg("SPLIT PARITION is not "
```
Comment (danielgustafsson) [inline src/backend/commands/tablecmds.c]: s/PARITION/PARTITION/

**Pull request merged**